klippa/src/lib.rs

```diff
@@ -7,67 +7,216 @@ mod parsing_util;
+
+            let mut vec_cap = input_gids.len() + input_unicodes.len();
+            vec_cap = vec_cap.min(cmap_unicodes.len());
+            self.codepoint_to_glyph.reserve(vec_cap);
+            self.unicode_to_new_gid_list.reserve(vec_cap);
+            //TODO: possible micro-optimize: set iteration over ranges next_range()? getting ranges is faster in Harfbuzz int set
+            for cp in cmap_unicodes.iter() {
+                match unicode_gid_map.get(&cp) {
+                    Some(gid) => {
+                        if !input_gids.contains(*gid) && !input_unicodes.contains(cp) {
+                            continue;
+                        }
+                        self.codepoint_to_glyph.insert(cp, *gid);
+                        self.unicode_to_new_gid_list.push((cp, *gid));
+                    }
+                    None => {
+                        continue;
+                    }
+                }
+            }
+
+            /* Add gids which where requested, but not mapped in cmap */
+            //TODO: possible micro-optimize: set iteration over ranges next_range()? getting ranges is faster in Harfbuzz int set
+            for gid in input_gids.iter() {
+                if gid.to_u32() >= self.font_num_glyphs as u32 {
+                    break;
+                }
+                self.glyphset_gsub.insert(gid);
             }
         }
+        self.glyphset_gsub
+            .extend(self.unicode_to_new_gid_list.iter().map(|t| t.1));
+        self.unicodes
+            .extend(self.unicode_to_new_gid_list.iter().map(|t| t.0));
+    }
+
+    pub fn populate_gids_to_retain(&mut self, font: &FontRef) {
+        //not-def
+        self.glyphset_gsub.insert(GlyphId::from(0));
+
+        //glyph closure for cmap
+        let cmap = font.cmap().expect("Error reading cmap table");
+        cmap.closure_glyphs(&self.unicodes, &mut self.glyphset_gsub);
+        remove_invalid_gids(&mut self.glyphset_gsub, self.font_num_glyphs);
+
+        //skip glyph closure for MATH table, it's not supported yet
+        //glyph closure for COLR
+        //TODO: colr_closure;
+        //remove_invalid_gids(glyphset_colr, font_num_glyphs);
+
+        /* Populate a full set of glyphs to retain by adding all referenced composite glyphs. */
+        let loca = font.loca(None).expect("Error reading loca table");
+        let glyf = font.glyf().expect("Error reading glyf table");
+        let operation_count =
+            self.glyphset_gsub.len() * (MAX_COMPOSITE_OPERATIONS_PER_GLYPH as usize);
+        for gid in self.glyphset_gsub.iter() {
+            glyf_closure_glyphs(
+                &loca,
+                &glyf,
+                gid,
+                &mut self.glyphset,
+                operation_count as i32,
+                0,
+            );
+        }
+        remove_invalid_gids(&mut self.glyphset, self.font_num_glyphs);
+    }
+}
+
+/// glyph closure for Composite glyphs in glyf table
+fn glyf_closure_glyphs(
```

Comment:
I would add a bit more documentation of how this is expected to work, in particular how it uses the return value.